### PR TITLE
util: add Inline storage helper-class and use it in a few places.

### PR DESCRIPTION
**Pull request by jmarantz**
Description: There were at least two different patterns for allocating a variable-sized array as part of a class:
 * placed-new of malloced block from static class method, referencing a private ctor/dtor
 * placed-new with operator-new and delete overrides.

This PR unifies those approaches using the second pattern, which was introduced by @brian-pane , providing a re-usable mix-in to standardize on a single pattern.

I think using this pattern will also simplify #6400 and make it easier to reason about by abstracting away the pattern of variable-size allocations.

While in here I threw in another usage of inlined arrays: for the symbol storage in symbol tables. It didn't appear to make a difference though in the overall memory usage though, but it did make an improvement in the memory consumed for a targeted test of memory consumption by symbol tables.

Risk Level: medium -- anything messing around with memory is at least medium risk
Testing: //test/... -- running asan now in parallel
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/common/utility.h [hunk 1]
```diff
@@ -632,4 +633,64 @@ template <class Value> struct TrieLookupTable {
   TrieEntry<Value> root_;
 };
 
+// Helper class for allocating classes with variable-sized inlined storage. See
+// InlineString below for usage.
+class InlineStorage : public NonCopyable {
```
Comment (danzh2010) [inline source/common/common/utility.h]: Though there is InlineString as an example of using this class, I think it could be better if you explicitly point out that the derived class should 
1. hide its constructor and have a public interface like c`create(size_t buffer_size, ...)`
2. the buffer member has to be the last field of the class.
3. take care of buffer alignment.
It would be great if we can enforce this in the code, but I couldn't think of a way to achieve that.

File: source/common/stats/symbol_table_impl.cc [hunk 2]
```diff
@@ -215,8 +215,8 @@ Symbol SymbolTableImpl::toSymbol(absl::string_view sv) {
     // a string_view pointing to it in the encode_map_. This allows us to only
     // store the string once. We use unique_ptr so copies are not made as
     // flat_hash_map moves values around.
-    auto str = std::make_unique<std::string>(std::string(sv));
-    auto encode_insert = encode_map_.insert({*str, SharedSymbol(next_symbol_)});
+    InlineStringPtr str = InlineString::create(sv);
```
Comment (danzh2010) [inline source/common/stats/symbol_table_impl.cc]: I'm not familiar with this class to say that we should switch to use hand-roll InlineString, considering std::string probably already uses short string optimization.

File: source/common/common/utility.h [hunk 3]
```diff
@@ -632,4 +633,64 @@ template <class Value> struct TrieLookupTable {
...
+   */
+  const char* data() const { return data_; }
+
+private:
+  // Constructor is declared private so that no one constructs one without the
+  // proper size allocation. to accommodate the variable-size buffer.
+  InlineString(const char* str, size_t size) : size_(size) { memcpy(data_, str, size); }
+
+  size_t size_;
+  char data_[];
```
Comment (wu-bin) [inline source/common/common/utility.h]: Note that depends on what data members are placed before data_, data_ may not point to the address after *this, instead it can point to a address in the padding portion of *this. It may not be what we want.

<pre>
Address (low to high):    ----------------------------------------------->>
                          |                         |                     |
                          |        sizeof(T)        |    buffer_size      |
                                                |
                                             data_
</pre>

Comment (jmarantz) [reply source/common/common/utility.h]: Understood. I'm not 100% sure if's worth the verbosity, but I added this comment above the declaration for InlineStorage:

```
// Perf note: The alignment will be correct and safe without further
// consideration as long as there are no casts. But for micro-optimization,
// consider this case:
//   struct MyStruct : public InlineStorage { uint64_t a_; uint16_t b_; uint8_t data_[]; };
// When compiled with a typical compiler on a 64-bit machine:
//   sizeof(MyStruct) == 16, because the compiler will round up from 10 for uint64_t alignment.
// So:
//   calling new (6) MyStruct() causes an allocation of 16+6=22, rounded up to 24 bytes.
// But data_ doesn't need 8-byte alignment, so it will wind up adjacent to the uint16_t.
//   ((char*) my_struct.data) - ((char*) &my_struct) == 10
// If we had instead declared data_[6], then the whole allocation would have fit in 16 bytes.
// Instead:
//   - the starting address of data will not be 8-byte aligned. This is not required
//     by the C++ standard for a uint8_t, but may be suboptimal on some processes.
//   - the 6 bytes of data will be at byte offsets 10 to 15, and bytes 16 to 23 will be
//     unused. This may be surprising to some users, and suboptimal in resource usage.
// One possible tweak is to declare data_ as a uint64_t[], or to use an alignas
// declaration. As always, micro-optimizations should be informed by
// microbenchmarks, showing the benefi.

```

File: source/common/stats/symbol_table_impl.cc (hunk 2 above)
Comment (jmarantz) [reply source/common/stats/symbol_table_impl.cc]: It does provide a slightly space win over std::string in the case where I applied it. I agree it's not a proven win in all cases and I added a caveat.

File: source/common/common/utility.h (hunk 1 above)
Comment (jmarantz) [reply source/common/common/utility.h]: Good point, added this class doc:

```
// Mix-in class for allocating classes with variable-sized inlined storage.
//
// Use this class by inheriting from it, ensuring that:
//  - The variable sized array is declared as VarType[] as the last
//    member variable of the class.
//  - YourType accurately describes the type that will be stored there,
//    to enable the compiler to perform correct alignment. No casting
//    should be needed.
//  - The class constructor is private, because you need to allocate the
//    class the placed new operator exposed in the protected section below.
//    Constructing the class directly will not provide space for the
//    variable-size data.
//  - You expose a public factory method that return a placement-new, e.g.
//      static YourClass* alloc(size_t num_elements, constructor_args...) {
//        new (num_elements * sizeof(VarType)) YourClass(constructor_args...);
//      }
//
// See InlineString below for an example usage.
```

File: source/common/common/utility.h (hunk 3 above)
Comment (jmarantz) [reply source/common/common/utility.h]: Also, RE custom allocator: as discussed f2f think it should be possible to templatize InlineStorage on an allocator class.

File: test/common/stats/symbol_table_impl_test.cc [hunk 4]
```diff
@@ -591,9 +591,9 @@ TEST(SymbolTableTest, Memory) {
     // libc++:
     // -------
     // string_mem_used:        6710912
-    // symbol_table_mem_used:  1743512 (3.85x)
+    // symbol_table_mem_used:  1742472 (3.85x)
     EXPECT_LT(symbol_table_mem_used, string_mem_used / 3);
-    EXPECT_LT(symbol_table_mem_used, 1750000);
+    EXPECT_EQ(symbol_table_mem_used, 1742472);
```
Comment (wu-bin) [inline test/common/stats/symbol_table_impl_test.cc]: nit: If the memory usage is deterministic, consider change the previous line to EXPECT_EQ as well?

Comment (jmarantz) [reply test/common/stats/symbol_table_impl_test.cc]: done.

File: source/common/common/utility.h [hunk 5]
```diff
@@ -632,4 +633,111 @@ template <class Value> struct TrieLookupTable {
...
+// When compiled with a typical compiler on a 64-bit machine:
+//   sizeof(MyStruct) == 16, because the compiler will round up from 10 for uint64_t alignment.
+// So:
+//   calling new (6) MyStruct() causes an allocation of 16+6=22, rounded up to 24 bytes.
+// But data_ doesn't need 8-byte alignment, so it will wind up adjacent to the uint16_t.
+//   ((char*) my_struct.data) - ((char*) &my_struct) == 10
+// If we had instead declared data_[6], then the whole allocation would have fit in 16 bytes.
+// Instead:
+//   - the starting address of data will not be 8-byte aligned. This is not required
+//     by the C++ standard for a uint8_t, but may be suboptimal on some processes.
```
Comment (mattklein123) [inline source/common/common/utility.h]: s/processes/processors ?

File: source/common/common/utility.h [hunk 6]
```diff
@@ -632,4 +633,111 @@ template <class Value> struct TrieLookupTable {
...
+   * @return a pointer to the first byte of the string.
+   */
+  const char* data() const { return data_; }
+
+private:
+  // Constructor is declared private so that no one constructs one without the
+  // proper size allocation. to accommodate the variable-size buffer.
+  InlineString(const char* str, size_t size) : size_(size) { memcpy(data_, str, size); }
+
+  size_t size_;
```
Comment (mattklein123) [inline source/common/common/utility.h]: If we are really going nutty here for size, can't this be uint32_t and save 4 bytes on 64-bit machines?

File: source/common/stats/heap_stat_data.h [hunk 7]
```diff
@@ -23,9 +23,10 @@ namespace Stats {
  * This structure is an alternate backing store for both CounterImpl and GaugeImpl. It is designed
  * so that it can be allocated efficiently from the heap on demand.
  */
-struct HeapStatData {
+struct HeapStatData : public InlineStorage {
 private:
   explicit HeapStatData(StatName stat_name) { stat_name.copyToStorage(symbol_storage_); }
+  ~HeapStatData() = default;
```
Comment (mattklein123) [inline source/common/stats/heap_stat_data.h]: Why is this needed out of curiosity?

File: source/common/stats/heap_stat_data.h [hunk 8]
```diff
@@ -23,9 +23,10 @@ namespace Stats {
  * This structure is an alternate backing store for both CounterImpl and GaugeImpl. It is designed
  * so that it can be allocated efficiently from the heap on demand.
  */
-struct HeapStatData {
+struct HeapStatData : public InlineStorage {
```
Comment (mattklein123) [inline source/common/stats/heap_stat_data.h]: It's a little magical to me that `symbol_storage_` is what is inline, given that we have a layer of indirection via another struct/object declaration. I'm not sure if there is any way to make this better but maybe a comment somewhere?
